File: scripts/message-examples.py
#!/usr/bin/env python3
"""Generate/check public wire examples. Reads only repository test fixtures."""
import argparse
import json
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1]
DEST = ROOT / 'examples/messages'
# Deliberately synthetic: a well-formed-looking digest is not settlement evidence.
DIGEST = '1' * 32


def render(value, depth=0):
    """Readable exact JSON: byte arrays stay on one line; objects are indented."""
    if isinstance(value, dict) and value:
        pad = '  ' * (depth + 1)
        return '{\n' + ',\n'.join(pad + json.dumps(k) + ': ' + render(v, depth + 1)
                                   for k, v in value.items()) + '\n' + '  ' * depth + '}'
    if isinstance(value, list) and any(isinstance(v, (dict, list)) for v in value):
        return '[\n' + ',\n'.join('  ' * (depth + 1) + render(v, depth + 1)
                                   for v in value) + '\n' + '  ' * depth + ']'
    return json.dumps(value, ensure_ascii=False)


def corpus():
    vector = json.loads((ROOT / 'fixtures/channel-signing-vectors.json').read_text())
    legacy = json.loads((ROOT / 'fixtures/signing-vectors.json').read_text())
    data = vector['statements']
    payload = lambda name: data[name]['payload']
    raw = lambda value: list(bytes.fromhex(value))
    signed = lambda name: {'payload': payload(name), 'signature': raw(data[name]['signature_hex'])}
    offer, terms, request = payload('offer'), payload('terms'), payload('request')
    channel = request['channel']
    request_id = bytes(request['request_id']).decode('ascii')
    cert = {'close': payload('close'),
            'buyer_signature': raw(data['close']['buyer_signature_hex']),
            'provider_signature': raw(data['close']['provider_signature_hex'])}
    files, entries = {}, []

    def add(profile, name, message, direction, summary, status='implemented', variant=False):
        path = f'{profile}/{name}.json'
        files[path] = message
        kind = message['message']['type'] if profile == 'channel' else message.get('method', message.get('kind'))
        entries.append({'file': path, 'profile': profile, 'type': kind, 'direction': direction,
                        'runtime_status': status, 'variant': variant, 'summary': summary})

    def frame(kind, fields, *, job=False, unfunded=False):
        return {'version': 1, 'method': vector['method'], 'buyer': offer['buyer'],
                'provider': offer['provider'], 'agreement_id': '0x' + '0' * 64 if unfunded else channel,
                'request_id': request_id if job else '', 'message': {'type': kind, **fields}}

    def message(kind, fields, direction, summary, **kwargs):
        status = kwargs.pop('status', 'implemented')
        add('channel', kind, frame(kind, fields, **kwargs), direction, summary, status)

    message('payment.offer_request', {'opening_nonce': offer['opening_nonce'],
            'result_hash': terms['result_hash'], 'jobs': terms['max_jobs'],
            'max_unit_price': terms['unit_price'], 'deposit': offer['deposit']},
            'buyer_to_provider', 'Request a ten-job ceiling; this walkthrough closes after one job.', unfunded=True)
    message('payment.offer', {'offer': signed('offer'), 'terms': terms},
            'provider_to_buyer', 'Signed provider terms before the buyer funds the channel.', unfunded=True)
    message('session.resume', {}, 'buyer_to_provider', 'Admit or resume the synthetic funded channel.')
    ready = {'phase': 'active', 'highest_credit': None, 'completed_jobs': '0',
             'transcript_hash': raw(vector['hashes']['transcript_start']),
             'certificate': None, 'redeemed_amount': '0'}
    message('session.ready', ready, 'provider_to_buyer', 'Freshly admitted; no credit has been issued.')
    message('payment.authorize', {'request': request, 'credit': signed('credit')},
            'buyer_to_provider', 'Preauthorize job-0001 for 1,000 MIST cumulative.', job=True)
    message('payment.acknowledge', {'ack': signed('ack')},
            'provider_to_buyer', 'Acknowledge durable credit receipt; this does not prove delivery or payment.', job=True)
    message('work.get', {'request_hash': payload('credit')['request_hash']},
            'buyer_to_provider', 'Retrieve the exact request already backed by a saved credit.', job=True)
    message('work.result', {'result': signed('result'), 'bytes': list((ROOT / 'fixtures/hello.txt').read_bytes())},
            'provider_to_buyer', 'Return the fixture bytes and their signed request/credit binding.', job=True)
    message('payment.close', {'close': {'payload': payload('close'),
            'signature': cert['buyer_signature']}}, 'buyer_to_provider', 'Propose an early close after one completed job.')
    message('payment.close_acknowledge', {'certificate': cert},
            'provider_to_buyer', 'Countersign the close after freezing; the buyer submits it separately to Sui.')
    message('payment.settlement', {'digest': DIGEST}, 'buyer_to_provider',
            'Specified post-submission hint. No runtime sender/handler exists; digest is synthetic.', status='specified_only')
    message('error', {'code': 'conflict', 'message': 'conflicting repeated credit'},
            'provider_to_buyer', 'Alternative failure branch: the same request conflicts with saved credit.', job=True)

    resumed = {**ready, 'highest_credit': signed('credit'), 'completed_jobs': '1',
               'transcript_hash': raw(vector['hashes']['transcript_final'])}
    for label, fields, summary in [
        ('resumed', resumed, 'Restart after one result, before closing; no funds redeemed yet.'),
        ('frozen', {**resumed, 'phase': 'frozen', 'certificate': cert},
         'Close certificate saved; onchain settlement has not yet been observed.'),
        ('closed', {**resumed, 'phase': 'closed', 'certificate': cert, 'redeemed_amount': '1000'},
         'Cooperative-close branch: 1,000 MIST paid and 11,000 refunded.'),
        ('refunded', {**resumed, 'phase': 'refunded', 'redeemed_amount': '1000'},
         'Alternative branch: unilateral 1,000-MIST redemption followed by residual expiry refund.'),
    ]:
        add('channel', f'session.ready.{label}', frame('session.ready', fields),
            'provider_to_buyer', summary, variant=True)

    quote, acceptance = legacy['quote'], legacy['acceptance']
    signed_quote = {'quote': quote, 'signature': raw(legacy['quote_signature'])}
    receipt = {'acceptance': acceptance, 'signature': raw(legacy['acceptance_signature'])}
    for name, value, direction, summary in [
        ('request.quote', {'method': 'quote', 'version': 1, 'buyer': quote['buyer'],
         'nonce': quote['nonce'], 'result_hash': quote['result_hash']}, 'buyer_to_provider', 'Request the fixture quote.'),
        ('response.quote', {'kind': 'quote', 'signed': signed_quote}, 'provider_to_buyer', 'Sign 1,000-MIST escrow terms.'),
        ('request.deliver', {'method': 'deliver', 'version': 1, 'escrow': acceptance['escrow']},
         'buyer_to_provider', 'Retrieve the funded fixture.'),
        ('response.result', {'kind': 'result', 'escrow': acceptance['escrow'],
         'bytes': list((ROOT / 'fixtures/hello.txt').read_bytes())}, 'provider_to_buyer', 'Return fixture bytes.'),
        ('request.accept', {'method': 'accept', 'version': 1, 'receipt': receipt},
         'buyer_to_provider', 'Sign acceptance after verifying the delivered fixture.'),
        ('response.settled', {'kind': 'settled', 'escrow': acceptance['escrow'], 'digest': DIGEST},
         'provider_to_buyer', 'Implemented settlement response; this illustrative digest is not a chain transaction.'),
        ('response.error', {'kind': 'error', 'code': 'request_failed', 'message': 'quote expired'},
         'provider_to_buyer', 'Alternative error branch; reconcile before taking economic action.'),
    ]:
        add('escrow', name, value, direction, summary)

    files['index.json'] = {'format_version': 1, 'synthetic': True,
        'notice': 'Offline examples from public signing vectors. Object IDs, network, times and digests are synthetic; do not fund the test keys.',
        'profiles': {'channel': {'alpn': 'm2m/payment/1', 'method': 'sui.channel.v1',
                     'schema': '../../schemas/channel-v1.schema.json'},
                     'escrow': {'alpn': 'm2m/fixture/1', 'schema': '../../schemas/exchange-v1.schema.json'}},
        'examples': entries}
    return files


def validate(files):
    try:
        from jsonschema import Draft202012Validator
    except ImportError as error:
        raise SystemExit('Schema checks require jsonschema; see examples/messages/README.md for setup.') from error
    schemas = {name: json.loads((ROOT / 'schemas' / path).read_text()) for name, path in
               [('channel', 'channel-v1.schema.json'), ('escrow', 'exchange-v1.schema.json')]}
    for schema in schemas.values():
        Draft202012Validator.check_schema(schema)
    validators = {name: Draft202012Validator(schema) for name, schema in schemas.items()}
    for item in files['index.json']['examples']:
        validators[item['profile']].validate(files[item['file']])
    # Read the declared variants from the schemas, not a second handwritten list.
    def tags(branches, name):
        return {branch['properties'][name]['const'] for branch in branches}
    entries = files['index.json']['examples']
    declared = tags(schemas['channel']['$defs']['Message']['oneOf'], 'type')
    actual = {item['type'] for item in entries if item['profile'] == 'channel' and not item['variant']}
    if actual != declared:
        raise SystemExit(f'Channel message coverage differs: missing={declared - actual}, extra={actual - declared}')
    for definition, direction, tag in [('Request', 'buyer_to_provider', 'method'), ('Response', 'provider_to_buyer', 'kind')]:
        declared = tags(schemas['escrow']['$defs'][definition]['oneOf'], tag)
        actual = {item['type'] for item in entries if item['profile'] == 'escrow' and item['direction'] == direction}
        if actual != declared:
            raise SystemExit(f'Escrow {definition} coverage differs: {actual ^ declared}')


def main():
    parser = argparse.ArgumentParser(description=__doc__)
    modes = parser.add_mutually_exclusive_group()
    modes.add_argument('--write', action='store_true', help='regenerate examples from public vectors')
    modes.add_argument('--check', action='store_true', help='check schemas, coverage, and exact regeneration (default)')
    args = parser.parse_args()
    files = corpus()
    validate(files)
    stale = []
    for name, data in files.items():
        path, expected = DEST / name, render(data) + '\n'
        if args.write:
            path.parent.mkdir(parents=True, exist_ok=True)
            path.write_text(expected)
        elif not path.exists() or path.read_text() != expected:
            stale.append(name)
    existing = {str(p.relative_to(DEST)) for folder in ['channel', 'escrow']
                for p in (DEST / folder).glob('*.json')}
    extra = existing - set(files)
    if stale or extra:
        raise SystemExit(f'Examples differ: stale={sorted(stale)}, unindexed={sorted(extra)}; regenerate with --write.')
    entries = files['index.json']['examples']
    channel_count = sum(e['profile'] == 'channel' and not e['variant'] for e in entries)
    escrow_count = sum(e['profile'] == 'escrow' and not e['variant'] for e in entries)
    print(f'{"Generated" if args.write else "Verified"} {len(files) - 1} examples '
          f'covering all {channel_count} channel and {escrow_count} escrow message shapes.')


if __name__ == '__main__':
    main()

File: tests/message_examples.rs
//! Offline wire examples must be readable by the implementation and form a
//! cryptographically consistent exchange. This does not establish chain state.
use anyhow::{Context, Result, ensure};
use m2m::{channel_protocol as cp, protocol as escrow};
use serde::{Serialize, de::DeserializeOwned};
use serde_json::Value;
use std::{fs, path::PathBuf};

fn read<T: DeserializeOwned>(file: &str) -> Result<T> {
    let path = PathBuf::from(env!("CARGO_MANIFEST_DIR"))
        .join("examples/messages")
        .join(file);
    serde_json::from_slice(&fs::read(path)?).with_context(|| file.to_owned())
}

fn channel(name: &str) -> Result<cp::Envelope> {
    read(&format!("channel/{name}.json"))
}

fn verify<T: Serialize>(statement: &cp::Signed<T>, key: &[u8]) -> Result<()> {
    escrow::verify(&statement.payload, &statement.signature, key)
}

fn certificate(cert: &cp::CloseCertificate, offer: &cp::Offer) -> Result<()> {
    escrow::verify(&cert.close, &cert.buyer_signature, &offer.buyer_key)?;
    escrow::verify(&cert.close, &cert.provider_signature, &offer.provider_key)
}

#[test]
fn every_channel_example_parses_and_all_embedded_signatures_verify() -> Result<()> {
    let cp::Message::Offer { offer, .. } = channel("payment.offer")?.message else {
        anyhow::bail!("expected offer")
    };
    let cp::Message::Close { close } = channel("payment.close")?.message else {
        anyhow::bail!("expected close")
    };
    let manifest: Value = read("index.json")?;
    for item in manifest["examples"].as_array().unwrap() {
        if item["profile"] != "channel" {
            continue;
        }
        let file = item["file"].as_str().unwrap();
        let envelope: cp::Envelope = read(file)?;
        envelope.validate().with_context(|| file.to_owned())?;
        assert_eq!(envelope.buyer, offer.buyer, "{file}");
        assert_eq!(envelope.provider, offer.provider, "{file}");
        let unfunded = matches!(
            envelope.message,
            cp::Message::OfferRequest { .. } | cp::Message::Offer { .. }
        );
        assert_eq!(
            envelope.agreement_id,
            if unfunded {
                escrow::Address::ZERO
            } else {
                close.channel
            },
            "{file}"
        );
        let check = match &envelope.message {
            cp::Message::Offer { offer, .. } => verify(offer, &offer.provider_key),
            cp::Message::Authorize { request, credit } => {
                assert_eq!(request.request_id, envelope.request_id.as_bytes());
                assert_eq!(request.channel, envelope.agreement_id);
                verify(credit, &offer.buyer_key)
            }
            cp::Message::Acknowledge { ack } => verify(ack, &offer.provider_key),
            cp::Message::Result { result, bytes } => {
                assert_eq!(escrow::hash(bytes), result.result_hash, "{file}");
                verify(result, &offer.provider_key)
            }
            cp::Message::Close { close } => verify(close, &offer.buyer_key),
            cp::Message::CloseAcknowledge { certificate: cert } => certificate(cert, &offer),
            cp::Message::Ready {
                highest_credit,
                certificate: cert,
                ..
            } => {
                if let Some(credit) = highest_credit {
                    verify(credit, &offer.buyer_key)?;
                }
                if let Some(cert) = cert {
                    certificate(cert, &offer)?;
                }
                Ok(())
            }
            cp::Message::OfferRequest { .. }
            | cp::Message::Resume { .. }
            | cp::Message::Get { .. }
            | cp::Message::Settlement { .. }
            | cp::Message::Error { .. } => Ok(()),
        };
        check.with_context(|| file.to_owned())?;
        assert_eq!(
            item["runtime_status"],
            if matches!(envelope.message, cp::Message::Settlement { .. }) {
                "specified_only"
            } else {
                "implemented"
            },
            "{file}"
        );
    }
    Ok(())
}

#[test]
fn channel_examples_bind_one_job_from_offer_through_recovery_and_close() -> Result<()> {
    let cp::Message::Offer { offer, terms } = channel("payment.offer")?.message else {
        anyhow::bail!("expected offer")
    };
    let cp::Message::OfferRequest {
        opening_nonce,
        result_hash,
        jobs,
        max_unit_price,
        deposit,
    } = channel("payment.offer_request")?.message
    else {
        anyhow::bail!("expected offer request")
    };
    assert_eq!(opening_nonce, offer.opening_nonce);
    assert_eq!(result_hash, terms.result_hash);
    assert_eq!(jobs, terms.max_jobs);
    assert!(terms.unit_price <= max_unit_price);
    assert_eq!(deposit, offer.deposit);
    assert_eq!(offer.terms_hash, cp::terms_hash(&terms)?);

    let cp::Message::Authorize { request, credit } = channel("payment.authorize")?.message else {
        anyhow::bail!("expected authorization")
    };
    let cp::Message::Acknowledge { ack } = channel("payment.acknowledge")?.message else {
        anyhow::bail!("expected acknowledgement")
    };
    let cp::Message::Get { request_hash } = channel("work.get")?.message else {
        anyhow::bail!("expected work retrieval")
    };
    let cp::Message::Result { result, bytes } = channel("work.result")?.message else {
        anyhow::bail!("expected work result")
    };
    let cp::Message::Close { close } = channel("payment.close")?.message else {
        anyhow::bail!("expected close")
    };
    let cp::Message::CloseAcknowledge { certificate: cert } =
        channel("payment.close_acknowledge")?.message
    else {
        anyhow::bail!("expected close certificate")
    };
    assert_eq!(cert.close, close.payload);
    assert_eq!(cert.buyer_signature, close.signature);

    // All signed statements share an economic domain, parties, offer and channel.
    let offer_json = serde_json::to_value(&offer.payload)?;
    for statement in [
        serde_json::to_value(&credit.payload)?,
        serde_json::to_value(&ack.payload)?,
        serde_json::to_value(&result.payload)?,
        serde_json::to_value(&close.payload)?,
    ] {
        for field in [
            "method",
            "version",
            "network",
            "package_id",
            "deployment",
            "buyer",
            "provider",
        ] {
            assert_eq!(statement[field], offer_json[field], "{field}");
        }
        assert_eq!(statement["channel"], serde_json::to_value(request.channel)?);
        assert_eq!(
            statement["offer_hash"],
            serde_json::to_value(cp::offer_hash(&offer)?)?
        );
    }
    for message in [
        "payment.authorize",
        "payment.acknowledge",
        "work.get",
        "work.result",
    ] {
        assert_eq!(channel(message)?.request_id.as_bytes(), request.request_id);
    }
    assert_eq!(request.terms_hash, offer.terms_hash);
    assert_eq!(request_hash, cp::request_hash(&request)?);
    assert_eq!(credit.request_hash, request_hash);
    assert_eq!(result.request_hash, request_hash);
    assert_eq!(ack.credit_hash, cp::credit_hash(&credit)?);
    assert_eq!(result.credit_hash, ack.credit_hash);
    assert_eq!(bytes, include_bytes!("../fixtures/hello.txt"));
    assert_eq!(result.result_hash, terms.result_hash);
    assert_eq!(request.request_sequence, 1);
    assert_eq!(credit.sequence, request.request_sequence);
    assert_eq!(ack.sequence, credit.sequence);
    assert_eq!(close.final_sequence, credit.sequence);
    assert_eq!(credit.cumulative_amount, terms.unit_price);
    assert_eq!(ack.cumulative_amount, credit.cumulative_amount);
    assert_eq!(close.final_amount, credit.cumulative_amount);
    assert_eq!(offer.deposit - close.final_amount, 11_000);

    let start = cp::transcript_start(request.channel, &offer)?;
    assert_eq!(credit.previous_transcript_hash, start);
    let end = cp::transcript_next(&start, &request, &credit, &ack, &result)?;
    assert_eq!(close.transcript_hash, end);
    for (name, expected_phase, completed, redeemed, has_cert) in [
        ("session.ready", "active", 0, 0, false),
        ("session.ready.resumed", "active", 1, 0, false),
        ("session.ready.frozen", "frozen", 1, 0, true),
        ("session.ready.closed", "closed", 1, 1000, true),
        ("session.ready.refunded", "refunded", 1, 1000, false),
    ] {
        let cp::Message::Ready {
            phase,
            highest_credit,
            completed_jobs,
            transcript_hash,
            certificate,
            redeemed_amount,
        } = channel(name)?.message
        else {
            anyhow::bail!("expected ready")
        };
        assert_eq!(phase, expected_phase, "{name}");
        assert_eq!(completed_jobs, completed, "{name}");
        assert_eq!(redeemed_amount, redeemed, "{name}");
        assert_eq!(certificate, has_cert.then(|| cert.clone()), "{name}");
        assert_eq!(
            highest_credit,
            (completed > 0).then(|| credit.clone()),
            "{name}"
        );
        assert_eq!(
            transcript_hash,
            if completed == 0 { &start } else { &end }.clone(),
            "{name}"
        );
    }
    Ok(())
}

#[test]
fn escrow_examples_bind_quote_delivery_and_signed_acceptance() -> Result<()> {
    let escrow::Response::Quote { signed } = read("escrow/response.quote.json")? else {
        anyhow::bail!("expected quote")
    };
    let quote = &signed.quote;
    quote.validate()?;
    escrow::verify(quote, &signed.signature, &quote.provider_key)?;
    let escrow::Request::Accept { receipt, .. } = read("escrow/request.accept.json")? else {
        anyhow::bail!("expected acceptance")
    };
    let acceptance = &receipt.acceptance;
    assert_eq!(*acceptance, quote.acceptance(acceptance.escrow)?);
    escrow::verify(acceptance, &receipt.signature, &quote.buyer_key)?;

    let manifest: Value = read("index.json")?;
    for item in manifest["examples"].as_array().unwrap() {
        if item["profile"] != "escrow" {
            continue;
        }
        let file = item["file"].as_str().unwrap();
        if item["direction"] == "buyer_to_provider" {
            let request: escrow::Request = read(file)?;
            request.validate()?;
            match request {
                escrow::Request::Quote {
                    buyer,
                    nonce,
                    result_hash,
                    ..
                } => {
                    assert_eq!(buyer, quote.buyer);
                    assert_eq!(nonce, quote.nonce);
                    assert_eq!(result_hash, quote.result_hash);
                }
                escrow::Request::Deliver { escrow, .. } => assert_eq!(escrow, acceptance.escrow),
                escrow::Request::Accept { .. } => {}
            }
        } else {
            match read::<escrow::Response>(file)? {
                escrow::Response::Result { escrow, bytes } => {
                    assert_eq!(escrow, acceptance.escrow);
                    assert_eq!(bytes, include_bytes!("../fixtures/hello.txt"));
                    assert_eq!(escrow::hash(&bytes), quote.result_hash);
                }
                escrow::Response::Settled { escrow, digest } => {
                    assert_eq!(escrow, acceptance.escrow);
                    ensure!(!digest.is_empty(), "missing illustrative digest");
                }
                escrow::Response::Quote { .. } | escrow::Response::Error { .. } => {}
            }
        }
    }
    Ok(())
}
